nat/traversal/pinger.go

```diff
@@ -158,7 +138,58 @@ func (p *Pinger) PingProvider(ip string, localPorts, remotePorts []int, proxyPor
 		conn.Close()
 	}
 
-	return localPort, remotePort, err
+	return localPort, remotePort, nil
+}
+
+// PingConsumer pings consumer with increasing TTL for every connection.
+func (p *Pinger) PingConsumer(ip string, localPorts, remotePorts []int, mappingKey string) {
+	conns, err := p.PingPeer(ip, localPorts, remotePorts, 2, 1)
+	if err != nil {
+		log.Error().Err(err).Msg("Failed to ping remote peer")
+		return
+	}
+
+	conn := conns[0]
+
+	p.eventPublisher.Publish(event.AppTopicTraversal, event.BuildSuccessfulEvent(StageName))
+	log.Info().Msgf("Ping received from: %s, sending OK from: %s", conn.RemoteAddr(), conn.LocalAddr())
+
+	go p.natProxy.handOff(mappingKey, conn)
+}
+
+// PingPeer pings remote peer with a defined configuration.
+// It returns n connections if possible or all available with error.
+func (p *Pinger) PingPeer(ip string, localPorts, remotePorts []int, initialTTL int, n int) (conns []*net.UDPConn, err error) {
+	log.Info().Msg("NAT pinging to remote peer")
+
+	stop := make(chan struct{})
+	defer close(stop)
+
+	ch, err := p.multiPing(ip, localPorts, remotePorts, initialTTL, stop)
+	if err != nil {
+		log.Err(err).Msg("Failed to ping remote peer")
+		return nil, err
+	}
+
+	for res := range ch {
+		if res.err != nil {
+			log.Warn().Err(res.err).Msg("One of the pings has error")
+		} else {
+			if err := ipv4.NewConn(res.conn).SetTTL(maxTTL); err != nil {
+				log.Warn().Err(res.err).Msg("Failed to set connection TTL")
+				continue
+			}
+
+			res.conn.Write([]byte("OK")) // notify peer that we are using this connection.
+
+			conns = append(conns, res.conn)
+			if len(conns) == n {
+				return conns, nil
```

Comment:
At this point multiPing can be stopped too.

Reply:
It will be stopped by the `defer close(stop)`